tools/codegen/src/ast/verified/mod.rs

```diff
@@ -139,7 +158,20 @@ impl TopDecl {
                 name: name.to_owned(),
                 size: 1,
             }),
-            _ => None,
+            _ => {
+                #[cfg(not(feature = "gen-primitive-types"))]
+                return None;
+
+                #[cfg(feature = "gen-primitive-types")]
```

Comment:
Two Issues:

- Even I said "Duplicated code are bad.", but this schema has to copy to this crate.

  Because when you compile this crate as a dependency which is pulled from `crates.io`, this file would not be existed.

  In the repository, add a CI check to make sure the two copies are the same.

- With "with-primitive-types" features, these lines will cause a lint error.

   Try adding `#[cfg(not(feature="with-primitive-types"))]` to `None`.

Reply:
Thank you.

1. Can I create a hard link for `primitive.mol` file cross `molecule-std` and `codegen`? Then the two `primitive.mol` file in different crate will always same.
2. 
> Try adding #[cfg(not(feature="with-primitive-types"))] to None.

I define the feature name as "gen-primitive-types" in `codegen`, and "with-primitive-types" in `molecule-std`, is it right?

What's the meaning of "adding #[cfg(not(feature="with-primitive-types"))] to None"?
Do I need to replace `#[cfg(not(feature="with-primitive-types"))]` to `None`?

Reply:
This way?
```rust
    fn new_primitive(name: &str) -> Option<Self> {
        match name {
            "byte" => Some(Primitive {
                name: name.to_owned(),
                size: 1,
            }),
            _ => {
                let mut ret = None;
                
                #[cfg(feature = "gen-primitive-types")]
                ret = primitive_types(&PathBuf::from("../../primitive_types.mol"))
                    .unwrap()
                    .iter()
                    .find(|primitive_decl| primitive_decl.name().eq(name))
                    .map(|primitive_decl| Primitive {
                        name: primitive_decl.name().to_string(),
                        size: primitive_decl.item_count(),
                    });
                
                ret
            }
        }
        .map(Self::Primitive)
    }
```

Reply:
I'm debugging another project which already reach deadline but I didn't finished

So I couldn't have time to compile this, I don't know what's your problem.


How about the following code:

```rust
#[cfg(not(feature = "gen-primitive-types"))]
None

#[cfg(feature = "gen-primitive-types")]
primitive_types(&PathBuf::from("../../primitive_types.mol")) ... ...
```

I remembered not clearly, but if `#[cfg(...)]` couldn't be used on statements, try this:

```rust
#[cfg(not(feature = "gen-primitive-types"))]
{ None }

#[cfg(feature = "gen-primitive-types")]
{
    primitive_types(&PathBuf::from("../../primitive_types.mol")) ... ...
}
```

Or try this:


```rust
#[cfg(not(feature = "gen-primitive-types"))]
a_func_return_none();

#[cfg(feature = "gen-primitive-types")]
a_func_return_primitive_types();
```